Should retry'd transfers reflect updated configuration?

Config before:

```
# oops, there's a typo:
directory /my_datta/
accept .*
```

If the problem in the config is fixed and the process is restarted:

```
directory /my_data/
accept .*
```

When sr3 starts up, failed messages will be recovered from the retry queue and their `new_...` fields will still have the original directory. This isn't wrong, but v2 works differently (I haven't looked at the code to prove it, but that's how I've seen it work). 

We're considering deleting the `new_` fields when recovering messages from retry queues.

If we don't change this, if a directory needs to be changed, the user would need to edit the retry queue file on disk. I assume it would be even harder to edit the contents of a retry queue stored in redis.

## Testing with v2:

Config:

```
broker amqps://dd.weather.gc.ca/
topicPrefix v02.post

expire 1h

subtopic bulletins.alphanumeric.*.*.CWAO.#

directory /does_not_exist/
accept .*UANT01_CWAO.*
reject .*
```

Logs where downloads fail, files get added to retry queue:

```
$ sr_subscribe foreground test_change_dir
2024-01-31 18:13:10,207 [WARNING] extended option topicPrefix = ['v02.post'] (unknown or not declared)
2024-01-31 18:13:10,207 [INFO] sr_subscribe test_change_dir start
2024-01-31 18:13:10,208 [INFO] log settings start for sr_subscribe (version: 2.23.08):
2024-01-31 18:13:10,208 [INFO]  inflight=.tmp events=create|delete|link|modify use_pika=False topic_prefix=v02.post dry_run=False
2024-01-31 18:13:10,208 [INFO]  inline=False events=create|delete|link|modify use_amqplib=False topic_prefix=v02.post
2024-01-31 18:13:10,208 [INFO]  suppress_duplicates=False basis=path retry_mode=True retry_ttl=300000ms tls_rigour=normal
2024-01-31 18:13:10,208 [INFO]  expire=300000ms reset=False message_ttl=None prefetch=25 accept_unmatch=False delete=False poll_without_vip=True
2024-01-31 18:13:10,208 [INFO]  heartbeat=300 sanity_log_dead=450 default_mode=000 default_mode_dir=775 default_mode_log=600 discard=False durable=True
2024-01-31 18:13:10,208 [INFO]  declare_queue=True declare_exchange=True bind_queue=True
2024-01-31 18:13:10,208 [INFO]  post_on_start=True preserve_mode=True preserve_time=True realpath_post=False base_dir=None follow_symlinks=False
2024-01-31 18:13:10,208 [INFO]  mirror=False flatten=/ realpath_post=False strip=0 base_dir=None report_back=True log_reject=False
2024-01-31 18:13:10,209 [INFO]  Plugins configured:
2024-01-31 18:13:10,209 [INFO]          do_download:
2024-01-31 18:13:10,209 [INFO]          do_get     :
2024-01-31 18:13:10,209 [INFO]          on_message:
2024-01-31 18:13:10,209 [INFO]          on_data:
2024-01-31 18:13:10,209 [INFO]          on_part:
2024-01-31 18:13:10,209 [INFO]          on_file: File_Log
2024-01-31 18:13:10,209 [INFO]          on_post: Post_Log
2024-01-31 18:13:10,210 [INFO]          on_heartbeat: Hb_Log Hb_Memory RETRY
2024-01-31 18:13:10,210 [INFO]          on_report:
2024-01-31 18:13:10,210 [INFO]          on_start:
2024-01-31 18:13:10,210 [INFO]          on_stop:
2024-01-31 18:13:10,210 [INFO] log_settings end.
2024-01-31 18:13:10,210 [INFO] sr_subscribe run
2024-01-31 18:13:10,211 [INFO] AMQP  broker(dd.weather.gc.ca) user(anonymous) vhost(/)
2024-01-31 18:13:10,211 [INFO] Using amqp module (AMQP 0-9-1)
2024-01-31 18:13:10,231 [ERROR] standard queue name based on: prefix=q_anonymous program_name=sr_subscribe exchange_split=False no=-1
2024-01-31 18:13:10,233 [INFO] Binding queue q_anonymous.sr_subscribe.test_change_dir.49229110.20130665 with key v02.post.bulletins.alphanumeric.*.*.CWAO.# from exchange xpublic on broker amqps://anonymous@dd.weather.gc.ca/
2024-01-31 18:13:10,243 [INFO] declared queue q_anonymous.sr_subscribe.test_change_dir.49229110.20130665 (anonymous@dd.weather.gc.ca)
2024-01-31 18:13:10,247 [INFO] reading from to anonymous@dd.weather.gc.ca, exchange: xpublic
2024-01-31 18:13:10,249 [INFO] report_back to anonymous@dd.weather.gc.ca, exchange: xs_anonymous
2024-01-31 18:13:10,249 [INFO] sr_retry on_heartbeat
2024-01-31 18:13:10,254 [INFO] No retry in list
2024-01-31 18:13:10,256 [INFO] sr_retry on_heartbeat elapse 0.007346
2024-01-31 18:13:30,520 [WARNING] making /does_not_exist: [Errno 13] Permission denied: '/does_not_exist'
2024-01-31 18:13:30,520 [ERROR] could not cd to directory /does_not_exist
2024-01-31 18:13:30,520 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:13:30,522 [WARNING] downloading again, attempt 2
2024-01-31 18:13:30,523 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:13:30,524 [WARNING] downloading again, attempt 3
2024-01-31 18:13:30,525 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:13:30,527 [INFO] appended to retry list file 20240131181315.523 https://dd4.weather.gc.ca /bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311813___53188
2024-01-31 18:13:30,531 [WARNING] making /does_not_exist: [Errno 13] Permission denied: '/does_not_exist'
2024-01-31 18:13:30,531 [ERROR] could not cd to directory /does_not_exist
2024-01-31 18:13:30,531 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:13:30,533 [WARNING] downloading again, attempt 2
2024-01-31 18:13:30,533 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:13:30,535 [WARNING] downloading again, attempt 3
2024-01-31 18:13:30,535 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:13:30,537 [INFO] appended to retry list file 20240131181328.723 https://dd4.weather.gc.ca /bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311813___02502
2024-01-31 18:13:35,664 [WARNING] making /does_not_exist: [Errno 13] Permission denied: '/does_not_exist'
2024-01-31 18:13:35,664 [ERROR] could not cd to directory /does_not_exist
2024-01-31 18:13:35,664 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:13:35,666 [WARNING] downloading again, attempt 2
2024-01-31 18:13:35,667 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:13:35,668 [WARNING] downloading again, attempt 3
2024-01-31 18:13:35,668 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:13:35,670 [INFO] appended to retry list file 20240131181332.60 https://dd5.weather.gc.ca /bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311813___06373
2024-01-31 18:13:45,921 [WARNING] making /does_not_exist: [Errno 13] Permission denied: '/does_not_exist'
2024-01-31 18:13:45,921 [ERROR] could not cd to directory /does_not_exist
2024-01-31 18:13:45,921 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:13:45,923 [WARNING] downloading again, attempt 2
2024-01-31 18:13:45,923 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:13:45,925 [WARNING] downloading again, attempt 3
2024-01-31 18:13:45,926 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:13:45,927 [INFO] appended to retry list file 20240131181337.964 https://dd4.weather.gc.ca /bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311813___37295
2024-01-31 18:14:25,404 [WARNING] making /does_not_exist: [Errno 13] Permission denied: '/does_not_exist'
2024-01-31 18:14:25,404 [ERROR] could not cd to directory /does_not_exist
2024-01-31 18:14:25,404 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:14:25,407 [WARNING] downloading again, attempt 2
2024-01-31 18:14:25,407 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:14:25,409 [WARNING] downloading again, attempt 3
2024-01-31 18:14:25,409 [ERROR] sr_subscribe/__do_download__: Could not download
2024-01-31 18:14:25,411 [INFO] appended to retry list file 20240131181413.110 https://dd4.weather.gc.ca /bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311814___47827
^C2024-01-31 18:14:28,973 [INFO] signal stop (SIGINT)
2024-01-31 18:14:28,973 [INFO] sr_subscribe stop
```

Retry disk queue file:

```
$ cat ~/.cache/sarra/subscribe/test_change_dir/sr_subscribe_test_change_dir_00.retry.new
["v02.post.bulletins.alphanumeric.20240131.UA.CWAO.18", {"atime": "20240131181315.523", "filename": "msg_ddsr-WXO-DD3_6b510db3c95732769433f66db5173290:fromPxatx:CWAO:UA:3:Direct:20240131181312", "from_cluster": "DDSR.CMC", "mtime": "20240131181315.523", "parts": "1,125,1,0,0", "source": "WXO-DD", "sum": "d,051bdad5c7c408423873bf265a361490", "sundew_extension": "fromPxatx:CWAO:UA:3:Direct:20240131181312", "to_clusters": "DDSR.CMC,DDI.CMC,CMC,SCIENCE,EDM"}, "20240131181315.523 https://dd4.weather.gc.ca /bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311813___53188"]
["v02.post.bulletins.alphanumeric.20240131.UA.CWAO.18", {"atime": "20240131181328.723", "filename": "msg_ddsr-WXO-DD3_841bea2477d3f732ffec114b048deffb:fromPxatx:CWAO:UA:3:Direct:20240131181327", "from_cluster": "DDSR.CMC", "mtime": "20240131181328.723", "parts": "1,124,1,0,0", "source": "WXO-DD", "sum": "d,29ef6a74e946d1ea3c41df13d293acc7", "sundew_extension": "fromPxatx:CWAO:UA:3:Direct:20240131181327", "to_clusters": "DDSR.CMC,DDI.CMC,CMC,SCIENCE,EDM"}, "20240131181328.723 https://dd4.weather.gc.ca /bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311813___02502"]
["v02.post.bulletins.alphanumeric.20240131.UA.CWAO.18", {"atime": "20240131181332.60", "filename": "msg_ddsr-WXO-DD3_8700d326c8d05cc9f46490a5a0e5d0ad:fromPxatx:CWAO:UA:3:Direct:20240131181331", "from_cluster": "DDSR.CMC", "mtime": "20240131181332.60", "parts": "1,117,1,0,0", "source": "WXO-DD", "sum": "d,d5bd4dfd78325b991b3e66a083444530", "sundew_extension": "fromPxatx:CWAO:UA:3:Direct:20240131181331", "to_clusters": "DDSR.CMC,DDI.CMC,CMC,SCIENCE,EDM"}, "20240131181332.60 https://dd5.weather.gc.ca /bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311813___06373"]
["v02.post.bulletins.alphanumeric.20240131.UA.CWAO.18", {"atime": "20240131181337.964", "filename": "msg_ddsr-WXO-DD3_4e1d0ddcd4158100c6bb18110029c8c8:fromPxatx:CWAO:UA:3:Direct:20240131181336", "from_cluster": "DDSR.CMC", "mtime": "20240131181337.964", "parts": "1,123,1,0,0", "source": "WXO-DD", "sum": "d,6a78de9a730e81a87751f79f50c23294", "sundew_extension": "fromPxatx:CWAO:UA:3:Direct:20240131181336", "to_clusters": "DDSR.CMC,DDI.CMC,CMC,SCIENCE,EDM"}, "20240131181337.964 https://dd4.weather.gc.ca /bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311813___37295"]
["v02.post.bulletins.alphanumeric.20240131.UA.CWAO.18", {"atime": "20240131181413.110", "filename": "msg_ddsr-WXO-DD3_1a758da1110ac2344c49e3a7e9a6a828:fromPxatx:CWAO:UA:3:Direct:20240131181411", "from_cluster": "DDSR.CMC", "mtime": "20240131181413.110", "parts": "1,118,1,0,0", "source": "WXO-DD", "sum": "d,4829fb74f6f5575b97dfa5d058888dad", "sundew_extension": "fromPxatx:CWAO:UA:3:Direct:20240131181411", "to_clusters": "DDSR.CMC,DDI.CMC,CMC,SCIENCE,EDM"}, "20240131181413.110 https://dd4.weather.gc.ca /bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311814___47827"]
```

Changed the config to use a directory that does exist, and only accept the files I know are in the retry queue:


```
broker amqps://dd.weather.gc.ca/
topicPrefix v02.post

expire 1h

subtopic bulletins.alphanumeric.*.*.CWAO.#

directory /tmp/exists/
#accept .*UANT01_CWAO.*
accept .*UANT01_CWAO_311813___53188.*
accept .*UANT01_CWAO_311813___02502.*
accept .*UANT01_CWAO_311813___06373.*
accept .*UANT01_CWAO_311813___37295.*
accept .*UANT01_CWAO_311814___47827.*
reject .*
```


Log showing v2 uses the updated directory to successfully download:

```
$ sr_subscribe foreground test_change_dir
2024-01-31 18:24:38,477 [WARNING] extended option topicPrefix = ['v02.post'] (unknown or not declared)
2024-01-31 18:24:38,477 [INFO] sr_subscribe test_change_dir start
2024-01-31 18:24:38,478 [INFO] log settings start for sr_subscribe (version: 2.23.08):
2024-01-31 18:24:38,478 [INFO]  inflight=.tmp events=create|delete|link|modify use_pika=False topic_prefix=v02.post dry_run=False
2024-01-31 18:24:38,478 [INFO]  inline=False events=create|delete|link|modify use_amqplib=False topic_prefix=v02.post
2024-01-31 18:24:38,478 [INFO]  suppress_duplicates=False basis=path retry_mode=True retry_ttl=3600000ms tls_rigour=normal
2024-01-31 18:24:38,478 [INFO]  expire=3600000ms reset=False message_ttl=None prefetch=25 accept_unmatch=False delete=False poll_without_vip=True
2024-01-31 18:24:38,478 [INFO]  heartbeat=300 sanity_log_dead=450 default_mode=000 default_mode_dir=775 default_mode_log=600 discard=False durable=True
2024-01-31 18:24:38,478 [INFO]  declare_queue=True declare_exchange=True bind_queue=True
2024-01-31 18:24:38,478 [INFO]  post_on_start=True preserve_mode=True preserve_time=True realpath_post=False base_dir=None follow_symlinks=False
2024-01-31 18:24:38,478 [INFO]  mirror=False flatten=/ realpath_post=False strip=0 base_dir=None report_back=True log_reject=False
2024-01-31 18:24:38,478 [INFO]  Plugins configured:
2024-01-31 18:24:38,478 [INFO]          do_download:
2024-01-31 18:24:38,478 [INFO]          do_get     :
2024-01-31 18:24:38,478 [INFO]          on_message:
2024-01-31 18:24:38,478 [INFO]          on_data:
2024-01-31 18:24:38,478 [INFO]          on_part:
2024-01-31 18:24:38,478 [INFO]          on_file: File_Log
2024-01-31 18:24:38,478 [INFO]          on_post: Post_Log
2024-01-31 18:24:38,478 [INFO]          on_heartbeat: Hb_Log Hb_Memory RETRY
2024-01-31 18:24:38,478 [INFO]          on_report:
2024-01-31 18:24:38,478 [INFO]          on_start:
2024-01-31 18:24:38,478 [INFO]          on_stop:
2024-01-31 18:24:38,478 [INFO] log_settings end.
2024-01-31 18:24:38,478 [INFO] sr_subscribe run
2024-01-31 18:24:38,479 [INFO] AMQP  broker(dd.weather.gc.ca) user(anonymous) vhost(/)
2024-01-31 18:24:38,479 [INFO] Using amqp module (AMQP 0-9-1)
2024-01-31 18:24:38,501 [INFO] Binding queue q_anonymous.sr_subscribe.test_change_dir.75389274.42625677 with key v02.post.bulletins.alphanumeric.*.*.CWAO.# from exchange xpublic on broker amqps://anonymous@dd.weather.gc.ca/
2024-01-31 18:24:38,502 [INFO] declared queue q_anonymous.sr_subscribe.test_change_dir.75389274.42625677 (anonymous@dd.weather.gc.ca)
2024-01-31 18:24:38,504 [INFO] reading from to anonymous@dd.weather.gc.ca, exchange: xpublic
2024-01-31 18:24:38,506 [INFO] report_back to anonymous@dd.weather.gc.ca, exchange: xs_anonymous
2024-01-31 18:24:38,506 [INFO] sr_retry on_heartbeat
2024-01-31 18:24:38,512 [INFO] Number of messages in retry list 5
2024-01-31 18:24:38,513 [INFO] sr_retry on_heartbeat elapse 0.007136
2024-01-31 18:24:38,518 [INFO] Retrying notice  20240131181315.523 https://dd4.weather.gc.ca/bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311813___53188
2024-01-31 18:24:38,534 [INFO] file_log downloaded to: /tmp/exists/UANT01_CWAO_311813___53188
2024-01-31 18:24:38,536 [INFO] Retrying notice  20240131181328.723 https://dd4.weather.gc.ca/bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311813___02502
2024-01-31 18:24:38,552 [INFO] file_log downloaded to: /tmp/exists/UANT01_CWAO_311813___02502
2024-01-31 18:24:38,553 [INFO] Retrying notice  20240131181332.60 https://dd5.weather.gc.ca/bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311813___06373
2024-01-31 18:24:38,588 [INFO] file_log downloaded to: /tmp/exists/UANT01_CWAO_311813___06373
2024-01-31 18:24:38,589 [INFO] Retrying notice  20240131181337.964 https://dd4.weather.gc.ca/bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311813___37295
2024-01-31 18:24:38,603 [INFO] file_log downloaded to: /tmp/exists/UANT01_CWAO_311813___37295
2024-01-31 18:24:38,604 [INFO] Retrying notice  20240131181413.110 https://dd4.weather.gc.ca/bulletins/alphanumeric/20240131/UA/CWAO/18/UANT01_CWAO_311814___47827
2024-01-31 18:24:38,616 [INFO] file_log downloaded to: /tmp/exists/UANT01_CWAO_311814___47827
^C2024-01-31 18:24:44,794 [INFO] signal stop (SIGINT)
2024-01-31 18:24:44,794 [INFO] sr_subscribe stop
```

--- a/docs/source/Reference/sr3_options.7.rst
+++ b/docs/source/Reference/sr3_options.7.rst
@@ -1528,6 +1528,19 @@ retryEmptyBeforeExit: <boolean> (default: False)
 Used for sr_insects flow tests. Prevents Sarracenia from exiting while there are messages remaining in the retry queue(s). By default, a post will cleanly exit once it has created and attempted to publish messages for all files in the specified directory. If any messages are not successfully published, they will be saved to disk to retry later. If a post is only run once, as in the flow tests, these messages will never be retried unless retryEmptyBeforeExit is set to True.
 
 
+retry_refilter <boolean> (default: False)
+-----------------------------------------
+
+The **retry_refilter** option alters how messages are reloaded when they are retrieved from
+a retry queue. The default way (value: False) is to repeat the transfer using exactly
+the same message as before. If **retry_refilter** is set (value: True) then all the
+message's calculated fields will be discarded, and the processing re-started from the gather
+phase (accept/reject processing will be repeated, destinations re-calculated.)
+
+The normal retry behaviour is use when the remote has had a failure, and need to 
+re-send later, while the retry_refilter option is used when recovering from configuration 
+file errors, and some messages had incorrect selection or destination criteria.
+
 retry_ttl <duration> (default: same as expire)
 ----------------------------------------------
 

--- a/docs/source/fr/Reference/sr3_options.7.rst
+++ b/docs/source/fr/Reference/sr3_options.7.rst
@@ -1503,6 +1503,21 @@ publiés avec succès, ils seront enregistrés sur le disque pour réessayer ult
 exécutée qu’une seule fois, comme dans les tests de flux, ces messages d'annonce ne seront jamais réessayés, sauf si
 retryEmptyBeforeExit est défini à True.
 
+retry_refilter <boolean> (par défaut : False)
+---------------------------------------------
+
+L'option **retry_refilter** modifie la façon dont les messages sont rechargés lorsqu'ils sont récupérés à partir 
+d'une file d'attente pour une nouvelle tentive de transfer (retry). La méthode par défaut (valeur : False) 
+consiste à répéter le transfert en utilisant exactement le même message que précédemment. Si **retry_refilter** 
+est défini (valeur : True), alors tous les Les champs calculés du message + seront supprimés et le 
+traitement redémarrera à partir de la phase *gather*  (le traitement d'acceptation/rejet sera 
+répété, les destinations recalculées.)
+
+Le comportement normal de nouvelle tentative (retry) est utilisé lorsque la destination a subit une 
+panne et doit renvoyer plus tard, tandis que l'option **retry_refilter** est utilisée lors de la récupération 
+de la configuration
+
+
 retry_ttl <duration> (défaut: identique à expire)
 -------------------------------------------------
 

--- a/sarracenia/config.py
+++ b/sarracenia/config.py
@@ -113,6 +113,7 @@ default_options = {
     'recursive' : True,
     'report': False,
     'retryEmptyBeforeExit': False,
+    'retry_refilter': False,
     'sanity_log_dead': 9999,
     'sourceFromExchange': False,
     'sundew_compat_regex_first_match_is_zero': False,
@@ -134,7 +135,7 @@ flag_options = [ 'acceptSizeWrong', 'acceptUnmatched', 'amqp_consumer', 'baseUrl
     'follow_symlinks', 'force_polling', 'inline', 'inlineOnly', 'inplace', 'logMetrics', 'logStdout', 'logReject', 'restore', \
     'messageDebugDump', 'mirror', 'timeCopy', 'notify_only', 'overwrite', 'post_on_start', \
     'permCopy', 'queueBind', 'queueDeclare', 'randomize', 'recursive', 'realpathPost', \
-    'reconnect', 'report', 'reset', 'retryEmptyBeforeExit', 'save', 'sundew_compat_regex_first_match_is_zero', \
+    'reconnect', 'report', 'reset', 'retry_refilter', 'retryEmptyBeforeExit', 'save', 'sundew_compat_regex_first_match_is_zero', \
     'sourceFromExchange', 'statehost', 'users', 'v2compatRenameDoublePost'
                 ]
 
@@ -2511,6 +2512,10 @@ class Config:
             help=
             'allows simultaneous use of multiple versions and types of messages'
         )
+        parser.add_argument('--retry_refilter',
+                            action='store_true',
+                            default=self.retry_refilter,
+                            help='repeat message processing when retrying transfers (default just resends as previous attempt.)')
         #FIXME: select/accept/reject in parser not implemented.
         parser.add_argument(
             '--select',

--- a/sarracenia/flow/__init__.py
+++ b/sarracenia/flow/__init__.py
@@ -1070,9 +1070,10 @@ class Flow:
             (len(self.worklist.incoming), len(self.worklist.rejected)))
 
     def gather(self) -> None:
+        so_far=0
         for p in self.plugins["gather"]:
             try:
-                new_incoming = p()
+                new_incoming = p(self.o.batch-so_far)
             except Exception as ex:
                 logger.error( f'flowCallback plugin {p} crashed: {ex}' )
                 logger.debug( "details:", exc_info=True )
@@ -1080,6 +1081,12 @@ class Flow:
 
             if len(new_incoming) > 0:
                 self.worklist.incoming.extend(new_incoming)
+                so_far += len(new_incoming) 
+
+            # if we gathered enough with a subset of plugins then return.
+            if so_far >= self.o.batch:
+                return
+
 
     def do(self) -> None:
 

--- a/sarracenia/flowcb/__init__.py
+++ b/sarracenia/flowcb/__init__.py
@@ -80,9 +80,9 @@ class FlowCB:
 
         Task: acknowledge messages from a gather source.
 
-    def gather(self) -> list::
+    def gather(self, messageCountMax) -> list::
 
-        Task: gather messages from a source... return a list of messages.
+        Task: gather messages from a source... return a list of messages 
 
               in a poll, gather is always called, regardless of vip posession.
               in all other components, gather is only called when in posession

--- a/sarracenia/flowcb/gather/file.py
+++ b/sarracenia/flowcb/gather/file.py
@@ -680,7 +680,7 @@ class File(FlowCB):
         self.queued_messages = []
         self.primed = False
 
-    def gather(self):
+    def gather(self, messageCountMax):
         """
            from sr_post.py/run 
 

--- a/sarracenia/flowcb/gather/message.py
+++ b/sarracenia/flowcb/gather/message.py
@@ -28,7 +28,7 @@ class Message(FlowCB):
         else:
             logger.critical('missing required broker specification')
 
-    def gather(self) -> list:
+    def gather(self, messageCountMax) -> list:
         """
            return a current list of messages.
         """

--- a/sarracenia/flowcb/log.py
+++ b/sarracenia/flowcb/log.py
@@ -60,9 +60,9 @@ class Log(FlowCB):
     def metricsReport(self):
         return { 'lagMax': self.lagMax, 'lagTotal':self.lagTotal, 'lagMessageCount':self.msgCount, 'rejectCount':self.rejectCount }
 
-    def gather(self):
+    def gather(self, messageCountMax):
         if set(['gather']) & self.o.logEvents:
-            logger.info('')
+            logger.info( f' messageCountMax: {messageCountMax} ')
 
         return []
 

--- a/sarracenia/flowcb/retry.py
+++ b/sarracenia/flowcb/retry.py
@@ -54,6 +54,11 @@ class Retry(FlowCB):
             return
 
         self.o.add_option( 'retry_driver', 'str', 'disk')
+
+        # retry_refilter False -- rety to send with existing processing.
+        # retry_refilter True -- re-ingest and re-apply processing (if it has changed.)
+        self.o.add_option( 'retry_refilter', 'flag', False)
+
         #queuedriver = os.getenv('SR3_QUEUEDRIVER', 'disk')
 
         if self.o.retry_driver == 'redis':
@@ -69,17 +74,53 @@ class Retry(FlowCB):
 
         logger.debug('logLevel=%s' % self.o.logLevel)
 
+
     def gather(self, qty) -> None:
         """
         If there are only a few new messages, get some from the download retry queue and put them into
         `worklist.incoming`.
+
+        Do this in the gather() entry point if retry_refilter is True.
+
         """
-        if not features['retry']['present'] :
+        if not features['retry']['present'] or not self.o.retry_refilter:
             return []
 
         if qty <= 0: return []
 
-        return self.download_retry.get(qty)
+        message_list = self.download_retry.get(qty)
+
+        # eliminate calculated values so it is refiltered from scratch.
+        for m in message_list:
+             for k in m:
+                 if k in m['_deleteOnPost'] or k.startswith('new_'):
+                     del m[k]
+             del m['_deleteOnPost']
+
+        return message_list
+
+
+    def after_accept(self, worklist) -> None:
+        """
+        If there are only a few new messages, get some from the download retry queue and put them into
+        `worklist.incoming`.
+
+        Do this in the after_accept() entry point if retry_refilter is False.
+
+        """
+        if not features['retry']['present'] or self.o.retry_refilter:
+            return
+
+        qty = (self.o.batch / 2) - len(worklist.incoming)
+        #logger.info('qty: %d len(worklist.incoming) %d' % ( qty, len(worklist.incoming) ) )
+
+        if qty <= 0: return
+
+        mlist = self.download_retry.get(qty)
+
+        #logger.debug("loading from %s: qty=%d ... got: %d " % (self.download_retry_name, qty, len(mlist)))
+        if len(mlist) > 0:
+            worklist.incoming.extend(mlist)
 
     def after_work(self, worklist) -> None:
         """

--- a/sarracenia/flowcb/retry.py
+++ b/sarracenia/flowcb/retry.py
@@ -69,24 +69,17 @@ class Retry(FlowCB):
 
         logger.debug('logLevel=%s' % self.o.logLevel)
 
-    def after_accept(self, worklist) -> None:
+    def gather(self, qty) -> None:
         """
         If there are only a few new messages, get some from the download retry queue and put them into
         `worklist.incoming`.
         """
         if not features['retry']['present'] :
-            return
+            return []
 
-        qty = (self.o.batch / 2) - len(worklist.incoming)
-        #logger.info('qty: %d len(worklist.incoming) %d' % ( qty, len(worklist.incoming) ) )
+        if qty <= 0: return []
 
-        if qty <= 0: return
-
-        mlist = self.download_retry.get(qty)
-
-        #logger.debug("loading from %s: qty=%d ... got: %d " % (self.download_retry_name, qty, len(mlist)))
-        if len(mlist) > 0:
-            worklist.incoming.extend(mlist)
+        return self.download_retry.get(qty)
 
     def after_work(self, worklist) -> None:
         """

--- a/sarracenia/flowcb/run.py
+++ b/sarracenia/flowcb/run.py
@@ -68,7 +68,7 @@ class Run(FlowCB):
             logger.error("subprocess.run failed err={}".format(err))
             logger.debug("Exception details:", exc_info=True)
 
-    def gather(self):
+    def gather(self, messageCountMax):
         """
            FIXME: this does not make sense. need to figure out how to get the 
            messages back from the script, perhaps using a json file reader?

--- a/sarracenia/flowcb/scheduled/__init__.py
+++ b/sarracenia/flowcb/scheduled/__init__.py
@@ -87,7 +87,7 @@ class Scheduled(FlowCB):
         now=datetime.datetime.fromtimestamp(time.time(),datetime.timezone.utc)
         self.update_appointments(now)
 
-    def gather(self):
+    def gather(self,messageCountMax):
 
         # for next expected post
         self.wait_until_next()

--- a/sarracenia/flowcb/scheduled/wiski.py
+++ b/sarracenia/flowcb/scheduled/wiski.py
@@ -131,7 +131,7 @@ class Wiski(Scheduled):
         
         return submit_tokenization_request
 
-    def gather(self): # placeholder
+    def gather(self,messageCountMax): # placeholder
         
         messages=[]
 
@@ -216,5 +216,5 @@ if __name__ == '__main__':
     logging.basicConfig(level=logging.DEBUG)
 
     me = Wiski(flow.o)
-    me.gather()
+    me.gather(flow.o.batch)
 

